icu4j/main/classes/core/src/com/ibm/icu/impl/ICUBinary.java

```diff
@@ -480,38 +685,35 @@ public static ByteBuffer getRequiredData(String itemPath) {
      */
     private static ByteBuffer getData(ClassLoader loader, String resourceName,
             String itemPath, boolean required) {
-        ByteBuffer bytes = getDataFromFile(itemPath);
-        if (bytes != null) {
-            return bytes;
-        }
-        if (loader == null) {
-            loader = ClassLoaderUtil.getClassLoader(ICUData.class);
-        }
-        if (resourceName == null) {
-            resourceName = ICUData.ICU_BASE_NAME + '/' + itemPath;
-        }
-        ByteBuffer buffer = null;
-        try {
-            @SuppressWarnings("resource")  // Closed by getByteBufferFromInputStreamAndCloseStream().
-            InputStream is = ICUData.getStream(loader, resourceName, required);
-            if (is == null) {
-                return null;
-            }
-            buffer = getByteBufferFromInputStreamAndCloseStream(is);
-        } catch (IOException e) {
-            throw new ICUUncheckedIOException(e);
-        }
-        return buffer;
+
+        return getData(loader, resourceName, itemPath, required, DataFilesCategory.NORMAL);
     }
 
-    private static ByteBuffer getDataFromFile(String itemPath) {
-        for (DataFile dataFile : icuDataFiles) {
-            ByteBuffer data = dataFile.getData(itemPath);
-            if (data != null) {
-                return data;
-            }
+    /**
+     * Loads an ICU binary data file and returns it as a ByteBuffer.
+     * The buffer contents is normally read-only, but its position etc. can be modified.
+     *
+     * @param loader Used for loader.getResourceAsStream() unless the data is found elsewhere.
+     * @param resourceName Resource name for use with the loader.
+     * @param itemPath Relative ICU data item path, for example "root.res" or "coll/ucadata.icu".
+     * @param required If the resource cannot be found,
+     *        this method returns null (!required) or throws an exception (required).
+     * @param dataFilesCategory Whether the data files parent chain is loaded as NORMAL, OVERRIDE, etc.
+     * @return The data as a read-only ByteBuffer,
+     *         or null if required==false and the resource could not be found.
+     * @throws MissingResourceException if required==true and the resource could not be found
+     */
+    public static ByteBuffer getData(ClassLoader loader, String resourceName,
+            String itemPath, boolean required, DataFilesCategory dataFilesCategory) {
```

Comment:
Idea: Instead of passing enum DataFilesCategory and dispatching on it, you could take a DataFiles object and call .getData() directly on that. It turns this function into a one-liner... and then you can inline it into its call sites.

What do you think?

Reply:
I would like a lot -- manual dispatch (ex: as I currently have) is a complexity smell -- but how would that desired change work with the code in `ICUResourceBundleReader.ReaderCache.createInstance()`?